### DISCUSSION TITLE
[react] Cannot compile Hooks FAQ example because RefObject.current is readonly

### DISCUSSION THREAD

**Discussion by Tomcc** [DefinitelyTyped/DefinitelyTyped#58512] [Issues with a @types package] [answered]
The [React Hooks FAQ](https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily) contains an example about how to correctly manage a long lived object:
```js
function Image(props) {
  const ref = useRef(null);

  // ✅ IntersectionObserver is created lazily once
  function getObserver() {
    if (ref.current === null) {

      // ERROR: TS2540: Cannot assign to 'current' because it is a read-only property.
      ref.current = new IntersectionObserver(onIntersect);
    }
    return ref.current;
  }

  // When you need it, call getObserver()
  // ...
}
```

This doesn't compile for me (with `"@types/react": "~17.0.37"`) because `RefObject.current` is `readonly.`

I'm not a React expert, but it looks like it's actually allowed to assign to it, at least in some cases where the ref isn't a true DOM reference.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
Thanks for the question. It is indeed a hook that often raises questions regarding typing it properly.

The intended pattern is using ` const ref = useRef<IntersectionObserver|null>(null);`: 
```tsx
import {  useRef } from 'react'

function Image({onIntersect}: any) {
  const ref = useRef<IntersectionObserver|null>(null);

  // ✅ IntersectionObserver is created lazily once
  function getObserver() {
    if (ref.current === null) {

      ref.current = new IntersectionObserver(onIntersect);
    }
    return ref.current;
  }

  // When you need it, call getObserver()
  // ...
}
```

[Playground Link](https://www.typescriptlang.org/play?ssl=4&ssc=55&pln=4&pc=3#code/JYWwDg9gTgLgBAbznArgZwKYCUMDM4C+cuUEIcA5FBgIYDGMFAUE7igHYPATtwCSIGgHMMACgQ8+7GBiiYGBAFxwa7AJ4BKRE2R0eaeNXwBeVJhy4APFJlyMXHgHkARpigA3WQB92KADZ+AHyivgEaANwsyAD00XCAoOT80rLyMNzsLm6eUHDAaHB01DQyACZwfjQAXsB+anA8dBg6xBwOvCIwmbLZoloIzcjA+KJGAHR0KFDU0nDGc3Chfn1RyKuxcACiWFiOWMoAKgDKAEwArAAsAAzKAMKq7BDwNGhowEK8MBCUE1MY0hRwZz2GjoDC5eB5FRwIolAC0PFqcDApDAshgalGA1WYx+03gpnYGAA7klbKl0l0PLJRJJknYGBEsQQsdQYJNeDjJnjIshmc11gB1AAWfzgaggKAWGAwZWAMAANAUaAE4B1KT0NPy4qMdUwCEA)

**Comment by Tomcc** [comment]
Thanks, that worked! 

I'm still a bit confused though.
My code looked like this:

```ts
    const ref = React.useRef<T>(null);
    const getSingleton = () => {
        if (ref.current === null) {
            ref.current = create();
        }
        return ref.current;
    };
    ...
```

Now that I think about it, shouldn't that be an error immediately (when passing `null` to `useRef`) rather than later when assigning to `.current`? After all, the type of the ref is `T`, not `T | null`, so null shouldn't be a valid value?

The current type signatures allow the creation of a `RefObject<T>` that is null and readonly - AFAIK you never want that, so this only allows for mistakes?
